core/src/main/java/org/springframework/security/oauth2/server/authorization/jose/NimbusJwtEncoder.java

```diff
@@ -0,0 +1,232 @@
+				if (ALG.equals(name)) {
+					// skip as alg is already set
+				} else if (TYP.equals(name)) {
+					String typValue = headers.getType();
+					if (typValue != null) {
+						jwsHeaderBuilder = jwsHeaderBuilder.type(new JOSEObjectType(typValue));
+					}
+				} else if (CTY.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.contentType(headers.getHeaderAsString(CTY));
+				} else if (CRIT.equals(name)) {
+					List<String> critValues = headers.getHeaderAsStringList(CRIT);
+					if (critValues != null) {
+						jwsHeaderBuilder = jwsHeaderBuilder.criticalParams(new HashSet<>(critValues));
+					}
+				} else if (JKU.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.jwkURL(headers.getHeaderAsURI(JKU));
+				} else if (JWK.equals(name)) {
+					String jwkAsString = headers.getHeaderAsString(JWK);
+					if (jwkAsString != null) {
+						jwsHeaderBuilder = jwsHeaderBuilder.jwk(com.nimbusds.jose.jwk.JWK.parse(jwkAsString));
+					}
+				} else if (X5U.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.x509CertURL(headers.getHeaderAsURI(X5U));
+				} else if (X5T.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.x509CertThumbprint(
+							Base64URL.from(headers.getHeaderAsString(X5T)));
+				} else if (X5T256.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.x509CertSHA256Thumbprint(
+							Base64URL.from(headers.getHeaderAsString(X5T256)));
+				} else if (X5C.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.x509CertChain(
+							X509CertChainUtils.toBase64List(JSONArrayUtils.parse(headers.getHeaderAsString(X5C))));
+				} else if (KID.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.keyID(headers.getHeaderAsString(KID));
+				} else if (B64.equals(name)) {
+					jwsHeaderBuilder = jwsHeaderBuilder.base64URLEncodePayload(headers.getHeaderAsBoolean(B64));
+				} else {
+					jwsHeaderBuilder = jwsHeaderBuilder.customParam(name, headersMap.get(name));
+				}
+			}
+		} catch (Exception ex) {
+			throw new JwtException(String.format(ENCODING_ERROR_MESSAGE_TEMPLATE, ex.getMessage()), ex);
+		}
+		return jwsHeaderBuilder.build();
+	}
+
+	private JWTClaimsSet createJwtClaimsSet(JwtClaimsSet claimsSet) {
+		JWTClaimsSet.Builder jwtClaimsSetBuilder = new JWTClaimsSet.Builder();
+		claimsSet.getClaims().forEach(jwtClaimsSetBuilder::claim);
+		return jwtClaimsSetBuilder.build();
+	}
+
+	public static class KeyManagerJwtEncoderBuilder {
+		private KeyManager keyManager;
+		private JwsAlgorithm jwsAlgorithm;
+
+		public KeyManagerJwtEncoderBuilder(KeyManager keyManager) {
+			Assert.notNull(keyManager, "keyManager cannot be null");
+			this.keyManager = keyManager;
+			this.jwsAlgorithm = SignatureAlgorithm.RS256;
+		}
+
+		public KeyManagerJwtEncoderBuilder jwsAlgorithm(JwsAlgorithm jwsAlgorithm) {
+			Assert.notNull(jwsAlgorithm, "jwsAlgorithm cannot be null");
+			this.jwsAlgorithm = jwsAlgorithm;
+			return this;
+		}
+
+		public NimbusJwtEncoder build() {
+			Key key = keyManager.getActiveKey();
```

Comment:
Is the `build` method the right place to get the key? I'm thinking about key rotation.

It makes me wonder if the `KeyManager` should be in the `NimbusJwtEncoder` constructor.

Reply:
I missed this use-case. Yes, we do need to have `KeyManager` in the `NimbusJwtEncoder` and fetch the active key to create the signer before signing the JWT. I'll make the appropriate changes in the next draft.

Reply:
Do you think it would be worth trying to find a key from the manager that works with the selected algorithm?